File: Hash/1027. Longest Arithmetic Subsequence/1027. Longest Arithmetic Subsequence.java
class Solution {
    public int longestArithSeqLength(int[] A) {
        int n = A.length;
        Map[] dp = new Map[n];
        for (int i = 0; i < n; i++) {
            dp[i] = new HashMap<>();
        }
        
        int rets = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                long diff = (long)A[i] - (long)A[j];
                if (!dp[j].containsKey(diff)) dp[j].put(diff, 1);
                int temp = (int)dp[j].get(diff)+1;
                dp[i].put(diff, Math.max((int)dp[i].getOrDefault(diff, 1), temp));
                rets = Math.max(rets, (int)dp[i].get(diff));
            }
        }
        return rets;
    }
}

//dp[i][diff]

File: Hash/1027. Longest Arithmetic Subsequence/1027. Longest Arithmetic Subsequence.cpp
class Solution {
public:
    int longestArithSeqLength(vector<int>& A) {
        int n = A.size();
        int dp[n][1001];
        std::memset(dp, 0, sizeof dp);
        
        int rets = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                int diff = A[i] - A[j] + 500;
                if (dp[j][diff] == 0) dp[j][diff] = 1;
                dp[i][diff] = max(dp[i][diff], dp[j][diff]+1);
                rets = max(rets, dp[i][diff]);
            }
        }
        return rets;
    }
};
